File: manara/reverseList.js
const reverseListRecursive = (a, b) => {
  if (!b) return a;
  const next = b.next;
  b.next = a;
  return reverseListRecursive(b, next);
};

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let ans = head;
  if (head && head.next) {
    ans = reverseListRecursive(head, head.next);
    head.next = null;
  }

  return ans;
};

var reverseListIterative = function (head) {
  let [left, right] = [null, head];

  while (right) {
    // track next right node
    const nextRight = right.next;

    // do the reverse
    right.next = left;

    left = right;
    right = nextRight;
  }

  return left;
};

File: JavaWorkSpace/workspace/src/JavaPlayGround/ReverseList.java
package JavaPlayGround;

public class ReverseList {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null, current = head;

        while (current != null) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }

    public ListNode reverseListRec(ListNode head) {
        return rev(head, null);
    }

    public ListNode rev(ListNode node, ListNode pre) {
        if (node == null)
            return pre;
        ListNode temp = node.next;
        node.next = pre;
        return rev(temp, node);
    }

}
